xds/internal/balancer/weightedtarget/weightedtarget_test.go

```diff
@@ -0,0 +1,227 @@
+
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/grpc/attributes"
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/roundrobin"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/internal/hierarchy"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+	"google.golang.org/grpc/xds/internal/balancer/balancergroup"
+	"google.golang.org/grpc/xds/internal/testutils"
+)
+
+type testConfigBalancerBuilder struct {
+	balancer.Builder
+}
+
+func newTestConfigBalancerBuilder() *testConfigBalancerBuilder {
+	return &testConfigBalancerBuilder{
+		Builder: balancer.Get(roundrobin.Name),
+	}
+}
+
+func (t *testConfigBalancerBuilder) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+	rr := t.Builder.Build(cc, opts)
+	return &testConfigBalancer{
+		Balancer:   rr,
+		V2Balancer: rr.(balancer.V2Balancer),
+	}
+}
+
+const testConfigBalancerName = "test_config_balancer"
+
+func (t *testConfigBalancerBuilder) Name() string {
+	return testConfigBalancerName
+}
+
+type stringBalancerConfig struct {
+	serviceconfig.LoadBalancingConfig
+	s string
+}
+
+func (t *testConfigBalancerBuilder) ParseConfig(c json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+	// Return string without quotes.
+	return stringBalancerConfig{s: string(c[1 : len(c)-1])}, nil
+}
+
+// testConfigBalancer takes the balancer config string and use it as a backend
+// address.
+type testConfigBalancer struct {
+	balancer.Balancer
+	balancer.V2Balancer
+}
+
+func (b *testConfigBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
+	c, ok := s.BalancerConfig.(stringBalancerConfig)
+	if !ok {
+		return fmt.Errorf("unexpected balancer config with type %T", s.BalancerConfig)
+	}
+	oneMoreAddr := resolver.Address{Addr: c.s}
+	s.BalancerConfig = nil
+	s.ResolverState.Addresses = append(s.ResolverState.Addresses, oneMoreAddr)
+	return b.Balancer.(balancer.V2Balancer).UpdateClientConnState(s)
+}
+
+func (b *testConfigBalancer) Close() {
+	b.V2Balancer.Close()
+}
+
+var (
+	wtbBuilder          balancer.Builder
+	wtbParser           balancer.ConfigParser
+	testBackendAddrStrs []string
+)
+
+const testBackendAddrsCount = 12
+
+func init() {
+	balancer.Register(newTestConfigBalancerBuilder())
+	for i := 0; i < testBackendAddrsCount; i++ {
+		testBackendAddrStrs = append(testBackendAddrStrs, fmt.Sprintf("%d.%d.%d.%d:%d", i, i, i, i, i))
+	}
+	wtbBuilder = balancer.Get(weightedTargetName)
+	wtbParser = wtbBuilder.(balancer.ConfigParser)
+
+	balancergroup.DefaultSubBalancerCloseTimeout = time.Millisecond
+}
+
+// TestWeightedTarget covers the cases that a sub-balancer is added and a
+// sub-balancer is removed. It verifies that the addresses and balancer configs
+// are forwarded to the right sub-balancer.
+//
+// This test is intended to test the glue code in weighted_target. Most of the
+// functionality tests are covered by the balancer group tests.
+func TestWeightedTarget(t *testing.T) {
```

Comment:
This test is quite large and requires that the reader spend a bit of time to understand what is happening.
Could you split this up into simpler tests?
There is no test for the scenario where an existing locality's weight changes?
Given that we are only testing the glue code (as mentioned in the comment), why is it necessary to test the picker? I guess all we need to make sure is that the child policy (for which we have a test balancer here) receives the correct service config and addresses right.

Reply:
All features are already covered by balancergroup tests.
So this is more of an e2e test (but not really e2e, because the ClientConn is fake).